From 34b801ea6637aa210f0fa55527dd242c7903a3f3 Mon Sep 17 00:00:00 2001
From: neeshy <neeshy@tfwno.gf>
Date: Wed, 24 May 2023 20:30:56 -0400
Subject: [PATCH 2/4] su: fix setting argv0

argv0 was being set to the user's login shell even when -p was
specified. Only populate newargv once the shell is properly determined.
---
 su.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/su.c b/su.c
index bc7a94f..812aa43 100644
--- a/su.c
+++ b/su.c
@@ -76,9 +76,6 @@ main(int argc, char *argv[])
 
 	shell = pw->pw_shell[0] == '\0' ? "/bin/sh" : pw->pw_shell;
 	if (lflag) {
-		newargv[0] = shell;
-		newargv[1] = "-l";
-		newargv[2] = NULL;
 		term = getenv("TERM");
 		clearenv();
 		setenv("HOME", pw->pw_dir, 1);
@@ -88,9 +85,10 @@ main(int argc, char *argv[])
 		setenv("TERM", term ? term : "linux", 1);
 		if (chdir(pw->pw_dir) < 0)
 			eprintf("chdir %s:", pw->pw_dir);
-	} else {
 		newargv[0] = shell;
-		newargv[1] = NULL;
+		newargv[1] = "-l";
+		newargv[2] = NULL;
+	} else {
 		if (pflag) {
 			shell = getenv("SHELL");
 		} else {
@@ -101,6 +99,8 @@ main(int argc, char *argv[])
 				setenv("LOGNAME", pw->pw_name, 1);
 			}
 		}
+		newargv[0] = shell;
+		newargv[1] = NULL;
 	}
 	if (strcmp(pw->pw_name, "root") == 0)
 		setenv("PATH", ENV_SUPATH, 1);
-- 
2.43.0

